### ISSUE TITLE
Application getting crash when switching frequently between rooms.

### ISSUE THREAD

**Issue by onlymnsh** [RocketChat/Rocket.Chat.Android#1586]
## Description

Application getting crash when switching between rooms.

## Devices and Versions
Your Rocket.Chat.Android version: 2.5.1
Your Rocket.Chat Server version: 0.62.0

Mobile device model and OS version: Lava - Android 8.1.0

## Steps to reproduce
Switch frequently between Rooms  will produce app crash.
1.Open Room and Press back button before load message.
2.Open another room and Press back button before load messages.
3.Repeat These steps few times and it will make App Crash.

## Logs
08-07 14:50:18.356 16467-16467/chat.rocket.android.dev E/AndroidRuntime: FATAL EXCEPTION: main
    Process: chat.rocket.android.dev, PID: 16467
    kotlinx.coroutines.experimental.CompletionHandlerException: Exception in completion handler InvokeOnCancel[@81a450a] for CancellableContinuation(DispatchedContinuation[CommonPool, chat.rocket.core.internal.rest.RestClientKt$handleRestCall$1@3f4a17b]){Cancelled}@87e6b98
        at kotlinx.coroutines.experimental.AbstractContinuation.completeStateUpdate(AbstractContinuation.kt:340)
        at kotlinx.coroutines.experimental.AbstractContinuation.updateStateToFinal(AbstractContinuation.kt:318)
        at kotlinx.coroutines.experimental.AbstractContinuation.tryCancel(AbstractContinuation.kt:188)
        at kotlinx.coroutines.experimental.AbstractContinuation.cancel(AbstractContinuation.kt:109)
        at kotlinx.coroutines.experimental.ChildContinuation.invoke(JobSupport.kt:987)
        at kotlinx.coroutines.experimental.JobSupport.notifyCancellation(JobSupport.kt:1037)
        at kotlinx.coroutines.experimental.JobSupport.tryMakeCancelling(JobSupport.kt:527)
        at kotlinx.coroutines.experimental.JobSupport.makeCancelling(JobSupport.kt:502)
        at kotlinx.coroutines.experimental.JobSupport.cancel(JobSupport.kt:474)
        at kotlinx.coroutines.experimental.JobSupport.cancelChildrenInternal(JobSupport.kt:600)
        at kotlinx.coroutines.experimental.JobSupport.makeCompletingInternal(JobSupport.kt:585)
        at kotlinx.coroutines.experimental.JobSupport.makeCompleting$kotlinx_coroutines_core(JobSupport.kt:538)
        at kotlinx.coroutines.experimental.JobSupport.makeCompletingOnCancel(JobSupport.kt:532)
        at kotlinx.coroutines.experimental.JobSupport.cancel(JobSupport.kt:475)
        at kotlinx.coroutines.experimental.Job$DefaultImpls.cancel$default(Job.kt:161)
        at chat.rocket.android.core.lifecycle.CancelStrategy.onDestroy(CancelStrategy.kt:17)
        at chat.rocket.android.core.lifecycle.CancelStrategy_LifecycleAdapter.callMethods(CancelStrategy_LifecycleAdapter.java:25)
        at androidx.lifecycle.SingleGeneratedAdapterObserver.onStateChanged(SingleGeneratedAdapterObserver.java:35)
        at androidx.lifecycle.LifecycleRegistry$ObserverWithState.dispatchEvent(LifecycleRegistry.java:355)
        at androidx.lifecycle.LifecycleRegistry.backwardPass(LifecycleRegistry.java:309)
        at androidx.lifecycle.LifecycleRegistry.sync(LifecycleRegistry.java:328)
        at androidx.lifecycle.LifecycleRegistry.moveToState(LifecycleRegistry.java:138)
        at androidx.lifecycle.LifecycleRegistry.handleLifecycleEvent(LifecycleRegistry.java:124)
        at androidx.fragment.app.Fragment.performDestroy(Fragment.java:2682)
        at androidx.fragment.app.FragmentManagerImpl.moveToState(FragmentManager.java:1591)
        at androidx.fragment.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1784)
        at androidx.fragment.app.FragmentManagerImpl.moveToState(FragmentManager.java:1852)
        at androidx.fragment.app.FragmentManagerImpl.dispatchStateChange(FragmentManager.java:3269)
        at androidx.fragment.app.FragmentManagerImpl.dispatchDestroy(FragmentManager.java:3260)
        at androidx.fragment.app.FragmentController.dispatchDestroy(FragmentController.java:274)
        at androidx.fragment.app.FragmentActivity.onDestroy(FragmentActivity.java:409)
        at androidx.appcompat.app.AppCompatActivity.onDestroy(AppCompatActivity.java:210)
        at android.app.Activity.performDestroy(Activity.java:7253)
        at android.app.Instrumentation.callActivityOnDestroy(Instrumentation.java:1254)
        at android.app.ActivityThread.performDestroyActivity(ActivityThread.java:4419)
        at android.app.ActivityThread.handleDestroyActivity(ActivityThread.java:4450)
        at android.app.ActivityThread.-wrap5(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1673)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6545)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:857)
08-07 14:50:18.357 16467-16467/chat.rocket.android.dev E/AndroidRuntime: Caused by: java.lang.IllegalStateException: request.tag() must not be null
        at chat.rocket.core.internal.rest.RestClientKt$handleRequest$$inlined$suspendCancellableCoroutine$lambda$3.invoke(RestClient.kt:117)
        at chat.rocket.core.internal.rest.RestClientKt$handleRequest$$inlined$suspendCancellableCoroutine$lambda$3.invoke(Unknown Source:2)
        at kotlinx.coroutines.experimental.InvokeOnCancel.invoke(AbstractContinuation.kt:387)
        at kotlinx.coroutines.experimental.AbstractContinuation.completeStateUpdate(AbstractContinuation.kt:338)

**Comment by luckcoolla** [comment]
I see same issue

Your Rocket.Chat.Android version: v2.5.0
Your Rocket.Chat Server version: 0.64.0

Mobile device model and OS version: S8 - Android 8.0.0

```
E/AndroidRuntime: FATAL EXCEPTION: main
    kotlinx.coroutines.experimental.CompletionHandlerException: Exception in completion handler InvokeOnCancel[@3c862d3] for CancellableContinuation(DispatchedContinuation[CommonPool, chat.rocket.core.internal.rest.RestClientKt$handleRestCall$1@ca08e10]){Cancelled}@c058d09
        at kotlinx.coroutines.experimental.AbstractContinuation.completeStateUpdate(AbstractContinuation.kt:348)
        at kotlinx.coroutines.experimental.AbstractContinuation.updateStateToFinal(AbstractContinuation.kt:324)
        at kotlinx.coroutines.experimental.AbstractContinuation.tryCancel(AbstractContinuation.kt:200)
        at kotlinx.coroutines.experimental.AbstractContinuation.cancel(AbstractContinuation.kt:121)
        at kotlinx.coroutines.experimental.ChildContinuation.invoke(JobSupport.kt:974)
        at kotlinx.coroutines.experimental.JobSupport.notifyCancellation(JobSupport.kt:1024)
        at kotlinx.coroutines.experimental.JobSupport.tryMakeCancelling(JobSupport.kt:514)
        at kotlinx.coroutines.experimental.JobSupport.makeCancelling(JobSupport.kt:490)
        at kotlinx.coroutines.experimental.JobSupport.cancel(JobSupport.kt:462)
        at kotlinx.coroutines.experimental.AbstractCoroutine.cancel(AbstractCoroutine.kt:178)
        at kotlinx.coroutines.experimental.JobSupport.cancelChildrenInternal(JobSupport.kt:587)
        at kotlinx.coroutines.experimental.JobSupport.makeCompletingInternal(JobSupport.kt:572)
        at kotlinx.coroutines.experimental.JobSupport.makeCompleting$kotlinx_coroutines_core(JobSupport.kt:525)
        at kotlinx.coroutines.experimental.JobSupport.makeCompletingOnCancel(JobSupport.kt:519)
        at kotlinx.coroutines.experimental.JobSupport.cancel(JobSupport.kt:463)
        at kotlinx.coroutines.experimental.Job$DefaultImpls.cancel$default(Job.kt:173)
        at chat.rocket.android.core.lifecycle.CancelStrategy.onDestroy(CancelStrategy.kt:17)
        at chat.rocket.android.core.lifecycle.CancelStrategy_LifecycleAdapter.callMethods(CancelStrategy_LifecycleAdapter.java:25)
        at androidx.lifecycle.SingleGeneratedAdapterObserver.onStateChanged(SingleGeneratedAdapterObserver.java:35)
        at androidx.lifecycle.LifecycleRegistry$ObserverWithState.dispatchEvent(LifecycleRegistry.java:355)
        at androidx.lifecycle.LifecycleRegistry.backwardPass(LifecycleRegistry.java:309)
        at androidx.lifecycle.LifecycleRegistry.sync(LifecycleRegistry.java:328)
        at androidx.lifecycle.LifecycleRegistry.moveToState(LifecycleRegistry.java:138)
        at androidx.lifecycle.LifecycleRegistry.handleLifecycleEvent(LifecycleRegistry.java:124)
        at androidx.fragment.app.Fragment.performDestroy(Fragment.java:2689)
        at androidx.fragment.app.FragmentManagerImpl.moveToState(FragmentManager.java:1591)
        at androidx.fragment.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1784)
        at androidx.fragment.app.FragmentManagerImpl.moveToState(FragmentManager.java:1852)
        at androidx.fragment.app.FragmentManagerImpl.dispatchStateChange(FragmentManager.java:3269)
        at androidx.fragment.app.FragmentManagerImpl.dispatchDestroy(FragmentManager.java:3260)
        at androidx.fragment.app.FragmentController.dispatchDestroy(FragmentController.java:274)
        at androidx.fragment.app.FragmentActivity.onDestroy(FragmentActivity.java:419)
        at androidx.appcompat.app.AppCompatActivity.onDestroy(AppCompatActivity.java:210)
```

**Comment by luciofm** [comment]
Thanks for reporting, we will take a look on the issue.

**Comment by luckcoolla** [comment]
Is this a real issue cause?
```
E/AndroidRuntime:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1374)
     Caused by: java.lang.IllegalStateException: request.tag() must not be null
        at 
chat.rocket.core.internal.rest.RestClientKt$handleRequest$$inlined$suspendCancellableCoroutine$lambda$3.invoke(RestClient.kt:126)
        at chat.rocket.core.internal.rest.RestClientKt$handleRequest$$inlined$suspendCancellableCoroutine$lambda$3.invoke(Unknown Source:2)
        at kotlinx.coroutines.experimental.InvokeOnCancel.invoke(AbstractContinuation.kt:391)
        at kotlinx.coroutines.experimental.AbstractContinuation.completeStateUpdate(AbstractContinuation.kt:346)
```

Cann't imagine how the request tag could be null, here is Request.java constructor code:
![screen shot 2018-08-20 at 12 52 33 pm](https://user-images.githubusercontent.com/3755157/44333648-1038f880-a478-11e8-9360-98065ab41e8a.png)

---
Upd.: I've tried to add tags for the restClient requests(Kotlin SDK project), but the issue is still being reproduced. Plus additional check on the CancelStratery.kt doesn't help here
   ```
 @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    fun onDestroy() {
        if (!jobs.isCancelled) {
            jobs.cancel()
        }
    }
```
Any more ideas regarding a possible solution?

**Comment by LluisFelip** [comment]
I found the same issue:

Rocket.Chat.Android version: v2.5.1

`kotlinx.coroutines.experimental.CompletionHandlerException: Exception in completion handler InvokeOnCancel[@7a101b] for CancellableContinuation(DispatchedContinuation[CommonPool, chat.rocket.core.internal.rest.RestClientKt$handleRestCall$1@4c345b8]){Cancelled}@ab2ef91
        at kotlinx.coroutines.experimental.AbstractContinuation.completeStateUpdate(AbstractContinuation.kt:340)
        at kotlinx.coroutines.experimental.AbstractContinuation.updateStateToFinal(AbstractContinuation.kt:318)
        at kotlinx.coroutines.experimental.AbstractContinuation.tryCancel(AbstractContinuation.kt:188)
        at kotlinx.coroutines.experimental.AbstractContinuation.cancel(AbstractContinuation.kt:109)
        at kotlinx.coroutines.experimental.ChildContinuation.invoke(JobSupport.kt:987)
        at kotlinx.coroutines.experimental.JobSupport.notifyCancellation(JobSupport.kt:1037)
        at kotlinx.coroutines.experimental.JobSupport.tryMakeCancelling(JobSupport.kt:527)
        at kotlinx.coroutines.experimental.JobSupport.makeCancelling(JobSupport.kt:502)
        at kotlinx.coroutines.experimental.JobSupport.cancel(JobSupport.kt:474)
        at kotlinx.coroutines.experimental.JobSupport.cancelChildrenInternal(JobSupport.kt:600)
        at kotlinx.coroutines.experimental.JobSupport.makeCompletingInternal(JobSupport.kt:585)
        at kotlinx.coroutines.experimental.JobSupport.makeCompleting$kotlinx_coroutines_core(JobSupport.kt:538)
        at kotlinx.coroutines.experimental.JobSupport.makeCompletingOnCancel(JobSupport.kt:532)
        at kotlinx.coroutines.experimental.JobSupport.cancel(JobSupport.kt:475)
        at kotlinx.coroutines.experimental.Job$DefaultImpls.cancel$default(Job.kt:161)
        at chat.rocket.android.core.lifecycle.CancelStrategy.onDestroy(CancelStrategy.kt:17)
        at chat.rocket.android.core.lifecycle.CancelStrategy_LifecycleAdapter.callMethods(CancelStrategy_LifecycleAdapter.java:25)
        at androidx.lifecycle.SingleGeneratedAdapterObserver.onStateChanged(SingleGeneratedAdapterObserver.java:35)
        at androidx.lifecycle.LifecycleRegistry$ObserverWithState.dispatchEvent(LifecycleRegistry.java:355)
        at androidx.lifecycle.LifecycleRegistry.backwardPass(LifecycleRegistry.java:309)
        at androidx.lifecycle.LifecycleRegistry.sync(LifecycleRegistry.java:328)
        at androidx.lifecycle.LifecycleRegistry.moveToState(LifecycleRegistry.java:138)
        at androidx.lifecycle.LifecycleRegistry.handleLifecycleEvent(LifecycleRegistry.java:124)
        at androidx.fragment.app.Fragment.performDestroy(Fragment.java:2682)
        at androidx.fragment.app.FragmentManagerImpl.moveToState(FragmentManager.java:1591)
        at androidx.fragment.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1784)
        at androidx.fragment.app.FragmentManagerImpl.moveToState(FragmentManager.java:1852)
        at androidx.fragment.app.FragmentManagerImpl.dispatchStateChange(FragmentManager.java:3269)
        at androidx.fragment.app.FragmentManagerImpl.dispatchDestroy(FragmentManager.java:3260)
        at androidx.fragment.app.FragmentController.dispatchDestroy(FragmentController.java:274)
        at androidx.fragment.app.FragmentActivity.onDestroy(FragmentActivity.java:409)
        at androidx.appcompat.app.AppCompatActivity.onDestroy(AppCompatActivity.java:210)
        at android.app.Activity.performDestroy(Activity.java:7395)
        at android.app.Instrumentation.callActivityOnDestroy(Instrumentation.java:1306)
        at android.app.ActivityThread.performDestroyActivity(ActivityThread.java:4443)
        at android.app.ActivityThread.handleDestroyActivity(ActivityThread.java:4476)
        at android.app.servertransaction.DestroyActivityItem.execute(DestroyActivityItem.java:39)
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:145)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:70)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1808)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6669)
08-27 09:49:49.751 22345-22345/? E/AndroidRuntime:     at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
     Caused by: java.lang.IllegalStateException: request.tag() must not be null
        at chat.rocket.core.internal.rest.RestClientKt$handleRequest$$inlined$suspendCancellableCoroutine$lambda$3.invoke(RestClient.kt:117)
        at chat.rocket.core.internal.rest.RestClientKt$handleRequest$$inlined$suspendCancellableCoroutine$lambda$3.invoke(Unknown Source:2)
        at kotlinx.coroutines.experimental.InvokeOnCancel.invoke(AbstractContinuation.kt:387)
        at kotlinx.coroutines.experimental.AbstractContinuation.completeStateUpdate(AbstractContinuation.kt:338)
        	... 45 more`

**Comment by rafaelks** [comment]
This is already resolved in 3.0.
